File: src/main/java/com/jore/epoc/bo/accounting/Account.java
package com.jore.epoc.bo.accounting;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import com.jore.jpa.AbstractBusinessObject;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;

@Entity
public class Account extends AbstractBusinessObject {
    @ManyToOne(optional = false)
    private FinancialAccounting accounting;
    private String number;
    private BigDecimal startBalance = BigDecimal.ZERO;
    private String name;
    private AccountType accountType;
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "creditAccount", orphanRemoval = true)
    private List<Booking> creditBookings = new ArrayList<>();
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "debitAccount", orphanRemoval = true)
    private List<Booking> debitBookings = new ArrayList<>();

    public Account() {
    }

    // TODO disable public constructor
    public Account(AccountType accountType, String number, String name) {
        this.accountType = accountType;
        this.number = number;
        this.name = name;
    }

    public void credit(Booking booking) {
        booking.setCreditAccount(this);
        creditBookings.add(booking);
    }

    public void debit(Booking booking) {
        booking.setDebitAccount(this);
        debitBookings.add(booking);
    }

    public BigDecimal getBalance(LocalDate valueDate) {
        Objects.requireNonNull(valueDate, "Value date must not be null.");
        BigDecimal result = startBalance;
        result = result.add(debitBookings.stream().filter(booking -> !booking.getValueDate().isAfter(valueDate)).map(booking -> booking.getAmount()).reduce(BigDecimal.ZERO, BigDecimal::add));
        result = result.add(creditBookings.stream().filter(booking -> !booking.getValueDate().isAfter(valueDate)).map(booking -> booking.getAmount().negate()).reduce(BigDecimal.ZERO, BigDecimal::add));
        return accountType.equals(AccountType.BALANCE_SHEET) ? result : result.negate();
    }

    public String getName() {
        return name;
    }

    public String getNumber() {
        return number;
    }

    public BigDecimal getStartBalance() {
        return startBalance;
    }

    public AccountType getType() {
        return accountType;
    }

    public void setAccounting(FinancialAccounting financialAccounting) {
        accounting = financialAccounting;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setStartBalance(BigDecimal startBalance) {
        this.startBalance = Objects.requireNonNull(startBalance, "Start balance must not be null.");
    }
}

File: EpocServer/Sources/App/Models/Accounting/Account.swift
import Foundation

/// A general ledger account with debit and credit bookings.
/// Equivalent to `com.jore.epoc.bo.accounting.Account`.
public final class Account {
    public var id: Int64?
    public weak var accounting: FinancialAccounting?
    public let number: String
    public let name: String
    public let accountType: AccountType
    public var startBalance: Decimal = 0
    public var creditBookings: [Booking] = []
    public var debitBookings: [Booking] = []

    public init(accountType: AccountType, number: String, name: String) {
        self.accountType = accountType
        self.number = number
        self.name = name
    }

    public func credit(_ booking: Booking) {
        booking.creditAccount = self
        creditBookings.append(booking)
    }

    public func debit(_ booking: Booking) {
        booking.debitAccount = self
        debitBookings.append(booking)
    }

    /// Returns the balance of this account as of the given value date.
    /// For balance sheet accounts: debit - credit
    /// For income statement accounts: credit - debit (negated)
    public func getBalance(_ valueDate: Date) -> Decimal {
        let debitSum = debitBookings
            .filter { $0.valueDate <= valueDate }
            .reduce(Decimal(0)) { $0 + $1.amount }
        let creditSum = creditBookings
            .filter { $0.valueDate <= valueDate }
            .reduce(Decimal(0)) { $0 + $1.amount }
        let result = startBalance + debitSum - creditSum
        return accountType == .balanceSheet ? result : -result
    }
}
